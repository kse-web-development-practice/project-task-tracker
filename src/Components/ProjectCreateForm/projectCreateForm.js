import React, { useState } from 'react'
import styles from './projectCreateForm.module.css'
import { FormInput } from '../FormInput/formInput'
import { Button } from '../Button/button'
import { FormTextarea } from '../FormTextarea/formTextarea'

export const ProjectCreateForm = () => {
  const [name, setName] = useState('')
  const [description, setDescription] = useState('')
  const [error, setError] = useState('')

  const handleSubmit = (event) => {
    event.preventDefault()
    if (!name) {
      setError('Enter project name.')
    } else {
      // post func
      setError('')
    }
  }

  return (
    <form className={styles.projectCreateForm} onSubmit={handleSubmit}>
      {error && <div className="error">{error}</div>}
      <FormInput
        labelText="Project Name:"
        type=""
        value={name}
        onChange={(e) => setName(e.target.value)}
      />
      <FormTextarea
        labelText="Project Description:"
        value={description}
        onChange={(e) => setDescription(e.target.value)}
      />
      <Button isMain={true} type="submit">
        Create
      </Button>
    </form>
  )
}
