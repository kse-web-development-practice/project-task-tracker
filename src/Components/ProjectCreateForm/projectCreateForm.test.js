import React from 'react'
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'
import { ProjectCreateForm } from './projectCreateForm'
import { useNavigate } from 'react-router-dom'

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn()
}))

describe('ProjectCreateForm Component', () => {
  const renderProjectCreateForm = (createFunc) =>
    render(<ProjectCreateForm createFunc={createFunc} username="testUser" />)

  test('default snapshot', () => {
    const { asFragment } = renderProjectCreateForm()
    expect(asFragment()).toMatchSnapshot()
  })

  test('error snapshot', () => {
    const { asFragment } = renderProjectCreateForm()

    fireEvent.click(screen.getByText('Create'))

    expect(asFragment()).toMatchSnapshot()
  })

  test('valid submission', async () => {
    const createFunc = jest.fn().mockResolvedValue({})
    const mockNavigate = jest.fn()
    useNavigate.mockReturnValue(mockNavigate)

    renderProjectCreateForm(createFunc)

    act(() => {
      fireEvent.change(screen.getByLabelText(/Project Name:/i), {
        target: { value: 'New Project' }
      })
      fireEvent.change(screen.getByLabelText(/Project Description:/i), {
        target: { value: 'Project description' }
      })

      fireEvent.click(screen.getByText('Create'))
    })

    await waitFor(() => {
      expect(createFunc).toHaveBeenCalledWith(
        'New Project',
        'Project description',
        'Not started',
        'testUser'
      )
      expect(mockNavigate).toHaveBeenCalledWith('/')
    })
  })
})
