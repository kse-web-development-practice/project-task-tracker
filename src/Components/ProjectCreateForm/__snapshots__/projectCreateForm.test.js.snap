// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCreateForm Component default snapshot 1`] = `
<DocumentFragment>
  <form
    class="projectCreateForm"
  >
    <label
      class="formLabel"
    >
      Project Name:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Project Description:
      <textarea
        class="formTextarea"
      />
    </label>
    <button
      class="button main "
    >
      Create
    </button>
  </form>
</DocumentFragment>
`;

exports[`ProjectCreateForm Component error snapshot 1`] = `
<DocumentFragment>
  <form
    class="projectCreateForm"
  >
    <div
      class="error"
    >
      Enter project name.
    </div>
    <label
      class="formLabel"
    >
      Project Name:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Project Description:
      <textarea
        class="formTextarea"
      />
    </label>
    <button
      class="button main "
    >
      Create
    </button>
  </form>
</DocumentFragment>
`;
